lib/*: Misc trivial fixes & warning silencing

andreittr opened the pull request:
### Description of changes

This changeset contains miscellaneous small fixes found by following compiler warnings. First two commits are fixes for genuine issues caught, while the rest fix unlikely cornercases in some configurations.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

andreittr committed:
lib/uksparsebuf: Fix overclearing of memory

This change fixes a bug where uk_sparsebuf_memclear would clear the
entire slice, instead of only up to the desired length.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/ukfs-ramfs: Fix special file invalid modes

This change corrects an oversight in code creating special file nodes,
whereby callers could create invalid files by specifying inconsistent
file modes as arguments. File creation now takes the explicit mode
argument as authoritative, and ignores irrelevant fields in the target
argument.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/posix-vfs-fstab: Fix unused function warning

This change silences an unused function warning for when no special
filesystem mounts are selected during configuration.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/posix-fdio: Fix unused argument warning

This change silences an unused argument warning in the sendfile shim
code when vfscore is disabled.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/nolibc: Fix return local variable warning

This change fixes a warning in getcwd about possibly returning a local
variable's address. The warning is technically true, as the caller
could, by sheer luck, pass buf == tmp which would cause a strdup.
This fix plugs that cornercase with a temporary local, preserving the
documented logic of calling strdup only if !buf.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/ukrandom: Fix nonstring char array warning

This change silences a compiler warning about a string literal being
truncated to fit into a char array. The array in question is not
intended to be used as a string, thus we add the appropriate attribute
to silence the warning.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/ukallocstack: Fix unused var warning

This change replaces asserts with defined validation and error paths,
as signaling error to the caller is the correct thing to do instead of
proceeding with invalid assumptions.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

mogasergiu commented on lib/ukallocstack/stack.c:
```diff
@@ -159,7 +159,8 @@ static void *stack_memalign(struct uk_alloc *a, __sz align, __sz size)
 	 * configuration: alignment must not be bigger than PAGE_SIZE.
 	 */
-	UK_ASSERT(align <= PAGE_SIZE);
+	if (unlikely(align > PAGE_SIZE))
```
Well at this point should we also convert the above alignment assertions?

andreittr replied:
:thinking: Good point; `memalign` is a public API with _no requirements_ on what you can pass as `align`. Having it crash / do wrong things for inconvenient arguments is very poor and brittle behavior. Will change :+1:

andreittr committed:
include/uk: Add nonstring attribute

This change adds a shorthand for the "nonstring" attribute to compiler.h

Checkpatch-Ignore: PREFER_DEFINED_ATTRIBUTE_MACRO
Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr commented:
Pushed update addressing comment & removing commits that change `UK_BUG()` into `UK_CRASH()` as these require more thoughtful consideration.

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>
Approved-by: Sergiu Moga <sergiu@unikraft.io>
